packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,191 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { unreachableCase } from "@fluidframework/common-utils";
+import { LocalReferencePosition, PropertySet } from "@fluidframework/merge-tree";
+import { IntervalCollection, SequenceInterval } from "./intervalCollection";
+import { SharedString } from "./sharedString";
+
+const IntervalEventType = {
+	CHANGE: 0,
+	ADD: 1,
+	DELETE: 2,
+	PROPERTYCHANGED: 3,
+} as const;
+
+type IntervalEventType = typeof IntervalEventType[keyof typeof IntervalEventType];
+
+type IntervalRevertible =
+	| {
+			event: typeof IntervalEventType.CHANGE;
+			interval: SequenceInterval;
+			start: LocalReferencePosition;
+			end: LocalReferencePosition;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.ADD;
+			interval: SequenceInterval;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.DELETE;
+			interval: SequenceInterval;
+			start: LocalReferencePosition;
+			end: LocalReferencePosition;
+			collection: IntervalCollection<SequenceInterval>;
+	  }
+	| {
+			event: typeof IntervalEventType.PROPERTYCHANGED;
+			interval: SequenceInterval;
+			propertyDeltas: PropertySet;
+	  };
+
+type TypedRevertible<T extends IntervalRevertible["event"]> = IntervalRevertible & { event: T };
+
+export function appendLocalAddToRevertibles(
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	revertibles.push({
+		event: IntervalEventType.ADD,
+		interval,
+		collection,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalDeleteToRevertibles(
+	string: SharedString,
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of deleted interval
+	// just use positions on the interval?
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: interval.start,
```

Comment:
I'm not sure you can just use the existing local references and expect them to still be valid by the time the revert happens. (Also for changeInterval)